experimental/printer/dom/chunks.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dom
+
+// TODO: do we need some kind of Zero concept for Chunks (rather than just relying on nil checks)?
+
+// Chunks represents a group of [Chunk]s that are formatted together.
+//
+// To facilitate splitting logic in Chunks, we keep track of the last non whitespace-only
+// [Chunk] inserted.
+type Chunks struct {
+	chunks                 []*Chunk
+	lastNonWhitespaceChunk *Chunk
+}
+
+// NewChunks creates a new empty Chunks.
+func NewChunks() *Chunks {
+	return &Chunks{}
+}
+
+// Insert inserts the given [Chunk]s.
+// Empty/nil [Chunk]s are dropped.
+func (c *Chunks) Insert(chunks ...*Chunk) {
+	for _, chunk := range chunks {
+		if chunk != nil {
+			c.chunks = append(c.chunks, chunk)
+			if !chunk.whitespaceOnly {
+				c.lastNonWhitespaceChunk = chunk
+			}
+			if chunk.child != nil {
+				// If there is a child, then we must walk down the tree to get the last non whitespace-only [Chunk].
+				if chunk.child.LastNonWhitespaceOnlyChunk() != nil {
+					c.lastNonWhitespaceChunk = chunk.child.LastNonWhitespaceOnlyChunk()
+				}
+			}
+		}
+	}
+}
+
+// LastNonWhitespaceOnlyChunk is the last non whitespace-only [Chunk] in the group of chunks.
+//
+// TODO: would be useful not having to check nil when this is called with a Zero concept.
+// TODO: this is currently being called in dom_test for a small toy parser implementation,
+// however, I'm not sure we need to expose this. Otherwise, this is mostly used to handle
+// splitting behaviour on [Chunk].
```

Comment:
I'm not entirely sure this should be exposed. I have a test case where I am using this, but I also don't think that test case is implemented "correctly". This seems to me like it should largely be an internal concept.